Source/WebCore/style/values/will-change/StyleWillChange.cpp

```diff
@@ -0,0 +1,245 @@
+    default:
+        return false;
+    }
+}
+
+bool WillChangeAnimatableFeature::propertyTriggersCompositing(CSSPropertyID property)
+{
+    switch (property) {
+    case CSSPropertyOpacity:
+    case CSSPropertyFilter:
+    case CSSPropertyBackdropFilter:
+    case CSSPropertyWebkitBackdropFilter:
+        return true;
+    default:
+        return false;
+    }
+}
+
+bool WillChangeAnimatableFeature::propertyTriggersCompositingOnBoxesOnly(CSSPropertyID property)
+{
+    // Don't trigger for perspective and transform-style, because those
+    // only do compositing if they have a 3d-transformed descendant and
+    // we don't want to do compositing all the time.
+    // Similarly, we don't want -webkit-overflow-scrolling-touch to
+    // always composite if there's no scrollable overflow.
+    switch (property) {
+    case CSSPropertyScale:
+    case CSSPropertyRotate:
+    case CSSPropertyTranslate:
+    case CSSPropertyTransform:
+    case CSSPropertyOffsetPath:
+        return true;
+    default:
+        return false;
+    }
+}
+
+// MARK: WillChangeAnimatableFeatures::Data
+
+bool WillChangeAnimatableFeatures::Data::operator==(const WillChangeAnimatableFeatures::Data& other) const
+{
+    return m_animatableFeatures == other.m_animatableFeatures;
+}
+
+bool WillChangeAnimatableFeatures::Data::containsScrollPosition() const
+{
+    return std::ranges::any_of(m_animatableFeatures, [](auto& feature) { return feature.feature() == Feature::ScrollPosition; });
+}
+
+bool WillChangeAnimatableFeatures::Data::containsContents() const
+{
+    return std::ranges::any_of(m_animatableFeatures, [](auto& feature) { return feature.feature() == Feature::Contents; });
+}
+
+bool WillChangeAnimatableFeatures::Data::containsProperty(CSSPropertyID property) const
+{
+    return std::ranges::any_of(m_animatableFeatures, [&](auto& feature) { return feature.property() == property; });
+}
+
+bool WillChangeAnimatableFeatures::Data::createsContainingBlockForAbsolutelyPositioned(bool isRootElement) const
+{
+    return createsContainingBlockForOutOfFlowPositioned(isRootElement)
+        || containsProperty(CSSPropertyPosition);
+}
+
+bool WillChangeAnimatableFeatures::Data::createsContainingBlockForOutOfFlowPositioned(bool isRootElement) const
+{
+    return containsProperty(CSSPropertyPerspective)
+        // CSS transforms
+        || containsProperty(CSSPropertyTransform)
+        || containsProperty(CSSPropertyTransformStyle)
+        || containsProperty(CSSPropertyTranslate)
+        || containsProperty(CSSPropertyRotate)
+        || containsProperty(CSSPropertyScale)
+        || containsProperty(CSSPropertyOffsetPath)
+        // CSS containment
+        || containsProperty(CSSPropertyContain)
+        // CSS filter & backdrop-filter
+        || (containsProperty(CSSPropertyBackdropFilter) && !isRootElement)
+        || (containsProperty(CSSPropertyWebkitBackdropFilter) && !isRootElement)
+        || (containsProperty(CSSPropertyFilter) && !isRootElement);
```

Comment:
Is this OK for performance? Iterating 11 times seems possibly not good?

Reply:
Just moved from WillChangeData. A bit set could definitely help here, though unclear if it’s necessary.